File: swift/Sources/PrismKit/Artifact.swift
import Foundation

/// Typed view of a versioned Prism model artifact directory.
///
/// The artifact is the cross-platform release contract: a manifest listing
/// one lowered program per backend, the label schema, the subword tokenizer
/// definition, and the calibration provenance. Programs emit final calibrated
/// probabilities, so consumers decode with argmax, the 0.5 threshold for
/// multi-valued morphology features, and ``LemmaEditRule/apply(to:)``.
public struct PrismArtifact: Sendable {
    public let directory: URL
    public let manifest: ArtifactManifest
    public let labels: ArtifactLabels

    public init(contentsOf directory: URL) throws {
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase

        let manifestURL = directory.appendingPathComponent("manifest.json")
        guard let manifestData = try? Data(contentsOf: manifestURL) else {
            throw PrismError.missingArtifactFile("manifest.json")
        }
        let manifest = try decoder.decode(ArtifactManifest.self, from: manifestData)

        let labelsURL = directory.appendingPathComponent(manifest.labelsFile)
        guard let labelsData = try? Data(contentsOf: labelsURL) else {
            throw PrismError.missingArtifactFile(manifest.labelsFile)
        }
        let labels = try decoder.decode(ArtifactLabels.self, from: labelsData)

        self.directory = directory
        self.manifest = manifest
        self.labels = labels
    }

    /// The lowered program serving the requested compute device.
    ///
    /// `cpu` selects the XNNPACK program, `gpu` a CoreML or MPS program, and
    /// `automatic` prefers GPU when the artifact provides one and falls back
    /// to CPU otherwise.
    public func program(for device: ComputeDevice) throws -> ArtifactProgram {
        guard let largest = try programs(for: device).last else {
            throw PrismError.deviceUnavailable(device)
        }
        return largest
    }

    /// All programs serving the device, sorted by capacity (smallest first).
    ///
    /// Artifacts may ship several fixed-shape programs per backend; runtimes
    /// pick the smallest program a batch fits into, so short sentences never
    /// pay the padding cost of the largest shapes.
    public func programs(for device: ComputeDevice) throws -> [ArtifactProgram] {
        let gpuBackends: Set<String> = ["coreml", "mps"]
        let backendPrograms: [ArtifactProgram]
        switch device {
        case .cpu:
            backendPrograms = manifest.programs.filter { $0.backend == "xnnpack" }
        case .gpu:
            backendPrograms = manifest.programs.filter {
                gpuBackends.contains($0.backend)
            }
        case .automatic:
            if let programs = try? programs(for: .gpu), !programs.isEmpty {
                return programs
            }
            backendPrograms = manifest.programs.filter { $0.backend == "xnnpack" }
        }
        guard !backendPrograms.isEmpty else {
            throw PrismError.deviceUnavailable(device)
        }
        return backendPrograms.sorted {
            ($0.shapes.subwordCount, $0.shapes.tokenCount)
                < ($1.shapes.subwordCount, $1.shapes.tokenCount)
        }
    }
}

public struct ArtifactManifest: Decodable, Sendable {
    public let manifestFormatVersion: Int
    public let artifactName: String
    public let artifactVersion: String
    public let languageTags: [String]
    public let labelsFile: String
    public let vocabularyFile: String
    public let characterUnicodeNormalization: String
    public let tokenizer: TokenizerContract
    public let programs: [ArtifactProgram]
    /// Shared external tensor-data files (program-data separation); absent
    /// in artifacts whose programs embed their weights.
    public let dataFiles: [ArtifactDataFile]?
    public let calibrationFile: String?
}

public struct ArtifactDataFile: Decodable, Sendable {
    public let fileName: String
    public let sha256: String
    public let sizeBytes: Int
}

public struct TokenizerContract: Decodable, Sendable {
    public let fileName: String
    public let className: String
    public let paddingTokenId: Int
}

public struct ArtifactProgram: Decodable, Sendable {
    public let fileName: String
    public let format: String
    public let backend: String
    public let precision: String
    public let sha256: String
    public let sizeBytes: Int
    public let shapes: FixedShapes
    public let outputNames: [String]
    /// External tensor-data files required at load time; nil or empty when
    /// the weights live inside the program file.
    public let dataFiles: [String]?
}

public struct FixedShapes: Decodable, Sendable {
    public let batchSize: Int
    public let subwordCount: Int
    public let tokenCount: Int
    public let characterCount: Int?
}

public struct ArtifactLabels: Decodable, Sendable {
    public let labelsFormatVersion: Int
    public let schema: LabelSchema
    public let characterVocabulary: CharacterVocabulary?
    public let maximumCharacterCount: Int?
}

public struct LabelSchema: Decodable, Sendable {
    public let upos: UposSchema
    public let morphology: MorphologySchema
    public let lemmaRules: LemmaRuleSchema
}

public struct UposSchema: Decodable, Sendable {
    public let labels: [String]
}

public struct MorphologySchema: Decodable, Sendable {
    public let features: [MorphologyFeature]
}

public struct MorphologyFeature: Decodable, Sendable {
    public let name: String
    public let values: [String]
    public let allowsMultipleValues: Bool
}

public struct LemmaRuleSchema: Decodable, Sendable {
    public let rules: [LemmaEditRule]
}

/// A lemma as a reversible edit of the token form.
///
/// Mirrors the Python reference: removals and additions operate on
/// characters of the NFC-normalized token.
public struct LemmaEditRule: Decodable, Sendable, Equatable {
    public let prefixRemoval: Int
    public let suffixRemoval: Int
    public let prefixAddition: String
    public let suffixAddition: String

    public func apply(to token: String) throws -> String {
        let characters = Array(token)
        guard prefixRemoval + suffixRemoval <= characters.count else {
            throw PrismError.invalidLemmaRule
        }
        let unchanged = characters[prefixRemoval..<(characters.count - suffixRemoval)]
        return prefixAddition + String(unchanged) + suffixAddition
    }
}

public struct CharacterVocabulary: Decodable, Sendable {
    public let characters: [String]

    /// Character-to-ID lookup for the character-CNN inputs.
    public func identifiers() -> [String: Int] {
        var lookup: [String: Int] = [:]
        for (index, character) in characters.enumerated() {
            lookup[character] = index
        }
        return lookup
    }
}

File: cpp/src/artifact.cpp
#include "prism/artifact.h"

#include <algorithm>
#include <fstream>
#include <stdexcept>
#include <tuple>

#include <nlohmann/json.hpp>

namespace prism::artifact {

namespace {

nlohmann::json ParseJsonFile(const std::filesystem::path& path)
{
    std::ifstream file(path);
    if (!file) {
        throw std::runtime_error("Missing artifact file: " + path.string());
    }
    return nlohmann::json::parse(file);
}

// Splits UTF-8 text into codepoint-sized byte ranges; lemma edit rules count
// removals in codepoints, never in bytes.
std::vector<std::size_t> CodepointOffsets(const std::string& text)
{
    std::vector<std::size_t> offsets;
    for (std::size_t index = 0; index < text.size(); ++index) {
        // Continuation bytes are 0b10xxxxxx; every other byte starts a codepoint.
        if ((static_cast<unsigned char>(text[index]) & 0xC0U) != 0x80U) {
            offsets.push_back(index);
        }
    }
    offsets.push_back(text.size());
    return offsets;
}

} // namespace

std::optional<std::string> LemmaEditRule::Apply(const std::string& token) const
{
    const auto offsets = CodepointOffsets(token);
    const auto length = offsets.size() - 1;
    if (static_cast<std::size_t>(prefix_removal + suffix_removal) > length) {
        return std::nullopt;
    }
    const auto begin = offsets[static_cast<std::size_t>(prefix_removal)];
    const auto end = offsets[length - static_cast<std::size_t>(suffix_removal)];
    return prefix_addition + token.substr(begin, end - begin) + suffix_addition;
}

Artifact::Artifact(const std::filesystem::path& directory)
    : directory_(directory)
{
    const auto manifest = ParseJsonFile(directory / "manifest.json");

    // Consumers decide language support from these manifest values; fail
    // loudly instead of guessing from directory or artifact names.
    for (const auto* required : {"artifact_name", "artifact_version", "language_tags"}) {
        if (!manifest.contains(required)) {
            throw std::runtime_error(
                std::string("manifest.json misses required metadata: ") + required);
        }
    }
    name_ = manifest.at("artifact_name").get<std::string>();
    version_ = manifest.at("artifact_version").get<std::string>();
    if (!manifest.at("language_tags").is_array()) {
        throw std::runtime_error("manifest.json language_tags must be an array of strings.");
    }
    for (const auto& tag : manifest.at("language_tags")) {
        language_tags_.push_back(tag.get<std::string>());
    }

    // Optional: raw-text segmentation abbreviations. Absent in legacy
    // artifacts, in which case the runtime uses a built-in default.
    if (manifest.contains("segmentation")) {
        const auto& segmentation = manifest.at("segmentation");
        if (segmentation.contains("abbreviations")) {
            for (const auto& abbreviation : segmentation.at("abbreviations")) {
                segmentation_abbreviations_.push_back(abbreviation.get<std::string>());
            }
        }
    }

    const auto& tokenizer = manifest.at("tokenizer");
    tokenizer_.file_name = tokenizer.at("file_name").get<std::string>();
    tokenizer_.padding_token_id = tokenizer.at("padding_token_id").get<std::int64_t>();

    for (const auto& entry : manifest.at("programs")) {
        Program program;
        program.file_name = entry.at("file_name").get<std::string>();
        program.backend = entry.at("backend").get<std::string>();
        program.precision = entry.at("precision").get<std::string>();
        const auto& shapes = entry.at("shapes");
        program.shapes.batch_size = shapes.at("batch_size").get<int>();
        program.shapes.subword_count = shapes.at("subword_count").get<int>();
        program.shapes.token_count = shapes.at("token_count").get<int>();
        if (shapes.contains("character_count") && !shapes.at("character_count").is_null()) {
            program.shapes.character_count = shapes.at("character_count").get<int>();
        }
        if (entry.contains("data_files")) {
            for (const auto& data_file : entry.at("data_files")) {
                program.data_files.push_back(data_file.get<std::string>());
            }
        }
        // The C++ engine executes on the CPU; other backends belong to
        // platform-specific runtimes and are skipped here.
        if (program.backend == "xnnpack") {
            programs_.push_back(std::move(program));
        }
    }
    if (programs_.empty()) {
        throw std::runtime_error("Artifact provides no CPU (xnnpack) program.");
    }
    std::sort(programs_.begin(), programs_.end(), [](const auto& a, const auto& b) {
        return std::tuple(a.shapes.subword_count, a.shapes.token_count)
            < std::tuple(b.shapes.subword_count, b.shapes.token_count);
    });

    const auto labels_file = manifest.at("labels_file").get<std::string>();
    const auto labels = ParseJsonFile(directory / labels_file);
    const auto& schema = labels.at("schema");

    for (const auto& label : schema.at("upos").at("labels")) {
        labels_.upos_labels.push_back(label.get<std::string>());
    }
    for (const auto& entry : schema.at("morphology").at("features")) {
        MorphologyFeature feature;
        feature.name = entry.at("name").get<std::string>();
        for (const auto& value : entry.at("values")) {
            feature.values.push_back(value.get<std::string>());
        }
        feature.allows_multiple_values = entry.at("allows_multiple_values").get<bool>();
        labels_.features.push_back(std::move(feature));
    }
    for (const auto& entry : schema.at("lemma_rules").at("rules")) {
        LemmaEditRule rule;
        rule.prefix_removal = entry.at("prefix_removal").get<int>();
        rule.suffix_removal = entry.at("suffix_removal").get<int>();
        rule.prefix_addition = entry.at("prefix_addition").get<std::string>();
        rule.suffix_addition = entry.at("suffix_addition").get<std::string>();
        labels_.lemma_rules.push_back(std::move(rule));
    }
    for (const auto& character : labels.at("character_vocabulary").at("characters")) {
        labels_.character_vocabulary.push_back(character.get<std::string>());
    }
    labels_.maximum_character_count = labels.at("maximum_character_count").get<int>();
}

} // namespace prism::artifact
